postgres-persistence/src/main/java/com/netflix/conductor/postgres/util/PostgresQueueListener.java

```diff
@@ -0,0 +1,207 @@
+import org.postgresql.PGNotification;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.netflix.conductor.core.exception.NonTransientException;
+import com.netflix.conductor.postgres.config.PostgresProperties;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+public class PostgresQueueListener {
+
+    private PGConnection pgconn;
+
+    private Connection conn;
+
+    private DataSource dataSource;
+
+    private HashMap<String, QueueStats> queues;
+
+    private boolean connected = false;
+
+    private boolean connecting = false;
+
+    private long lastNotificationTime = 0;
+
+    private Integer stalePeriod;
+
+    protected final Logger logger = LoggerFactory.getLogger(getClass());
+
+    public PostgresQueueListener(DataSource dataSource, PostgresProperties properties) {
+        logger.info("Using experimental PostgresQueueListener");
+        this.dataSource = dataSource;
+        this.stalePeriod = properties.getExperimentalQueueNotifyStalePeriod();
+        connect();
+    }
+
+    public boolean hasMessagesReady(String queueName) {
+        checkUpToDate();
+        handleNotifications();
+        if (notificationIsStale() || !connected) {
+            connect();
+            return true;
+        }
+
+        QueueStats queueStats = queues.get(queueName);
+        if (queueStats == null) {
+            return false;
+        }
+
+        if (queueStats.getNextDelivery() > System.currentTimeMillis()) {
+            return false;
+        }
+
+        return true;
+    }
+
+    public Integer getSize(String queueName) {
+        checkUpToDate();
+        handleNotifications();
+        if (notificationIsStale() || !connected) {
+            connect();
+            return null;
+        }
+
+        QueueStats queueStats = queues.get(queueName);
+        if (queueStats == null) {
+            return 0;
+        }
+
+        return queueStats.getDepth();
+    }
+
+    private boolean notificationIsStale() {
+        return System.currentTimeMillis() - lastNotificationTime > this.stalePeriod;
+    }
+
+    private void connect() {
+        if (!connecting) {
+            connected = false;
+            connecting = true;
+            try {
+                this.conn = dataSource.getConnection();
+                this.pgconn = conn.unwrap(PGConnection.class);
+
+                try {
+                    boolean previousAutoCommitMode = conn.getAutoCommit();
+                    conn.setAutoCommit(true);
+                    try {
+                        conn.prepareStatement("LISTEN conductor_queue_state").execute();
+                    } catch (Throwable th) {
+                        conn.rollback();
+                        logger.info(th.getMessage());
+                    } finally {
+                        conn.setAutoCommit(previousAutoCommitMode);
+                        connected = true;
```

Comment:
If the listen failed, why do we mark the connection as connected?

Reply:
Good spot - I've moved this